### DISCUSSION TITLE
coraza consumes too much cpu resources, can we do performance optimization

### DISCUSSION THREAD

**Discussion by Axkea** [corazawaf/coraza#881] [Q&A]
Here are the flame charts I generated during my performance tests, which can be used as a reference for performance optimization
 https://drive.google.com/file/d/1965ETYFolv62OFfA6SqKKFfoT-bIc60j/view?usp=drive_link

**Comment by Axkea** [comment]
This is the top 10 for pprof （cpu）
(pprof) top
Showing nodes accounting for 16520ms, 46.35% of 35640ms total
Dropped 352 nodes (cum <= 178.20ms)
Showing top 10 nodes out of 137
      flat  flat%   sum%        cum   cum%
    3410ms  9.57%  9.57%     6270ms 17.59%  regexp.(*Regexp).tryBacktrack
    3180ms  8.92% 18.49%     5180ms 14.53%  regexp.(*machine).add
    2400ms  6.73% 25.22%     2400ms  6.73%  [envoy]
    1330ms  3.73% 28.96%     3300ms  9.26%  runtime.scanobject
    1250ms  3.51% 32.46%     2770ms  7.77%  regexp.(*machine).step
    1220ms  3.42% 35.89%    26570ms 74.55%  github.com/corazawaf/coraza/v3/internal/corazawaf.(*Rule).doEvaluate
    1200ms  3.37% 39.25%     1860ms  5.22%  runtime.findObject
     890ms  2.50% 41.75%      890ms  2.50%  regexp.(*bitState).shouldVisit (inline)
     830ms  2.33% 44.08%     3640ms 10.21%  runtime.mallocgc
     810ms  2.27% 46.35%      920ms  2.58%  regexp.(*bitState).push (inline)

**Comment by jcchavezs** [comment]
cc'ing @anuraaga as he worked in allocations, performance and GC.

**Reply by anuraaga** [reply]
This looks to mostly be showing regexp processing in the flame graph. It is known that Go's regexp is slow

https://github.com/golang/go/issues/26623

An alternative is to use coraza-wasilibs with re2

https://github.com/corazawaf/coraza-wasilibs

While the Wasm codepath has issues with concurrency until Wasm threads proposal reaches stage 4, it looks like you are using cgo already, in which case you can install `re2`, e.g. `apt-get install libre2-dev`, and use the `re2_cgo` build tag to activate the cgo path. I think it will improve your performance.
